src/charm.py

```diff
@@ -483,6 +550,31 @@ def _tracing_endpoint_info_grpc(self) -> str:
     def _log_level(self) -> str:
         return self.config["log_level"]
 
+    @property
+    def _internal_url(self) -> str:
+        """Return workload's internal URL. Used for ingress."""
+        return f"http://{socket.getfqdn()}:{ADMIN_UI_PORT}"
+
+    @property
+    def _domain_url(self) -> Optional[str]:
+        return self.ingress.url if self.ingress.is_ready() else self._internal_url
+
+    @property
+    def _oauth_client_config(self) -> ClientConfig:
+        c = ClientConfig(
+            os.path.join(self._domain_url, OAUTH_CALLBACK_PATH),
+            OAUTH_SCOPES,
+            OAUTH_GRANT_TYPES,
+        )
+        # Bootstrap the client config to have the client_id as an aud.
+        # TODO(nsklikas): Remove when the login-ui automatically adds it to the audience
+        # https://github.com/canonical/identity-platform-login-ui/issues/244
+        if hasattr(self, "oauth"):
```

Comment:
Just a question: why do we check the attribute here?

Reply:
This function is used to initialize the `oauth` attribute (https://github.com/canonical/identity-platform-admin-ui-operator/pull/11/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR147). So before the `oauth` attribute is initialized we can't do `self.oauth.get_provider_info`